### PR TITLE
Support current Google Maps URL formats (/place/, ?q=, short links) in GoogleMapsUrlFormat

### LINKED ISSUE #301: [maintainer] Support current Google Maps URL formats (/place/, ?q=, short links) in GoogleMapsUrlFormat
## Bug: GoogleMapsUrlFormat can't parse three current-day Google Maps URL shapes

**Source:** maintainer request (Christian, 2026-08-16): "Check if the GoogleMapsUrlFormat is still able to work on current Google Maps URLs." Goal: verify + fix format drift against live Google Maps, since the class's own test fixtures are all 2007–2015 vintage.

**Empirical verification (2026-08-16, live google.com):** multi-stop `/maps/dir/A/B/C/@lat,lng,zoom/data=...` directions links are unchanged since 2015 and still parse correctly (`isParseableUrl` → true, named waypoints extracted, matches existing `testParseNewGoogleMaps2014FromInput9`/`2015FromInput10` behaviour — captured live: `https://www.google.com/maps/dir/Hamburg/Hannover/M%C3%BCnchen/@50.7220765,5.277764,6z/data=...?entry=ttu&g_ep=...`). No regression there — leave `/dir/` handling untouched.

Three other current, common URL shapes are silently rejected or silently mis-parsed today (confirmed against `GoogleMapsUrlFormat.URL_PATTERN`/`isParseableUrl`/`parsePositions` as of this repo's current `main`):

1. **`/place/<name>/@lat,lng,zoom/data=...`** — single-location links (the result of searching a place and clicking Share, or "What's here"). Live capture: `https://www.google.com/maps/place/Berlin/@52.5068767,13.4247534,10z/data=!3m1!4b1!4m6!3m5!1s0x47a84e373f035901:0x42120465b5e3b70!8m2!3d52.5200066!4d13.404954!16zL20vMDE1NnE?entry=ttu&g_ep=...`. `isParseableUrl` requires `found.startsWith("?")` or `found.startsWith("/dir/")` — `/place/...` matches neither → **completely rejected**, not even attempted.
2. **`?q=<lat,lng>` or `?q=<name>`** (e.g. `https://maps.google.com/maps?q=52.520008,13.404954`) — the standard Android/iOS "share location" geo-intent fallback URL, still generated by countless non-Google apps/integrations. `isParseableUrl` passes (`found` starts with `"?"`), but `parsePositions(Map<String,List<String>>)` only ever reads `saddr`/`daddr` — the `q` parameter is silently ignored → **passes the parseable check but yields zero positions**, a silent data-loss bug, not a clean rejection.
3. **`https://maps.app.goo.gl/<id>`** (today's default Share-button output on both mobile and desktop) and legacy **`https://goo.gl/maps/<id>`** — short links that 30x-redirect to a full Maps URL. `URL_PATTERN` requires the literal substring `.google.` in the host, which neither short-link domain contains → **`findURL` returns null, format isn't even selected** for these at all.

Verified via a standalone regex simulation of `URL_PATTERN`/`isParseableUrl` against all three shapes plus the working `/dir/` case (results: `/dir/` → parseable=True, `/place/` → parseable=False, `?q=` → parseable=True but parses to empty, `maps.app.goo.gl` → found=None).

This spec is fully resolved (grilled 2026-08-16); implement without further questions.

### Scope
Fix all three gaps in one PR (same class, same regex/parsing mechanism, cohesive change):
- Recognize and parse `/place/<name>/...` single-location links, extracting both the place name and its exact coordinate.
- Recognize and parse `?q=<value>` links (plain coordinate pair or place name), when no `saddr`/`daddr` is present.
- Recognize `maps.app.goo.gl/...` and `goo.gl/maps/...` short links, resolve the redirect chain (bounded, max 5 hops) to the real Maps URL, then parse that URL with the (now-extended) existing logic.

**Out of scope:**
- Extracting exact coordinates for `/dir/` multi-waypoint named places — stays name-only, unchanged, exactly as today's `testParseNewGoogleMaps2014FromInput9`/`2015FromInput10` already assert. Do not touch `/dir/` parsing.
- Bare map-center URLs with no place, e.g. `.../maps/@52.52,13.40,15z` — carries no position semantically, correctly stays unparsed.
- Any user-facing preference to disable network lookups for short-link resolution — no other network-touching code in this codebase has one either (checked `DownloadExecutor`, `PositionAugmenter`), stay consistent and omit it.
- `isGoogleMapsProfileUrl` / `.url` bookmark-file handling (`GoogleMapsUrlFormatBookmarkTest`) — unaffected, do not change.

### Files
- `navigation-formats/src/main/java/slash/navigation/url/GoogleMapsUrlFormat.java` — main fix (see Behaviour below)
- `rest/src/main/java/slash/navigation/rest/HttpRequest.java` — expose redirect control (new protected method only)
- `rest/src/main/java/slash/navigation/rest/ReadRequest.java` — expose redirect control + `Location` header (new public methods only)
- `navigation-formats/src/test/java/slash/navigation/url/GoogleMapsUrlFormatTest.java` — new tests (see Testable seam below)

### Behaviour (decisions locked)

**A. `rest` module — new redirect-control API (house idiom: reuse this module's Apache HttpClient5 wrapper, not raw `HttpURLConnection`; no existing code in the repo does redirect resolution today, so this is new but must follow the existing wrapper's shape).**

In `HttpRequest.java`, add (next to the existing `disableContentCompression()`):
```java
protected void disableRedirectHandling() {
    clientBuilder.disableRedirectHandling();
}
```

In `ReadRequest.java`, add (next to the existing `getETag()`/`setCacheControlNoCache()` — reuse the already-imported `import static org.apache.hc.core5.http.HttpHeaders.*;`, which already provides `LOCATION`):
```java
public void disableRedirectHandling() {
    super.disableRedirectHandling();
}

public String getLocationHeader() throws IOException {
    return getHeader(LOCATION);
}
```
(`Head` and `Get` inherit both automatically — no changes needed in `Head.java`/`Get.java`.)

**B. `GoogleMapsUrlFormat.java` — short-link detection + bounded redirect resolution.**

Add a new pattern (next to `URL_PATTERN`):
```java
private static final Pattern SHORT_URL_PATTERN = Pattern.compile(".*(http[s]?://(?:maps\\.app\\.goo\\.gl|goo\\.gl/maps)/[^\\s]+).*");
private static final int MAX_REDIRECT_HOPS = 5;
```

In `internalFindUrl(String text)`, check `SHORT_URL_PATTERN` **before** the existing `URL_PATTERN` check (short-link domains never match `URL_PATTERN` anyway, but check order still matters for clarity/short-circuit):
```java
Matcher shortUrlMatcher = SHORT_URL_PATTERN.matcher(text);
if (shortUrlMatcher.matches())
    return "SHORTLINK:" + shortUrlMatcher.group(1);
```
(keep the existing bookmark-unwrap and `URL_PATTERN` logic below this, unchanged).

Update `isParseableUrl`:
```java
public boolean isParseableUrl(String url) {
    String found = internalFindUrl(url);
    return found != null && (found.startsWith("?") || found.startsWith("/dir/") ||
            found.startsWith("/place/") || found.startsWith("SHORTLINK:"));
}
```

Add a testable, pure redirect-following seam (package-private, no network inside — network is injected via the functional interface):
```java
interface RedirectResolver {
    String resolveLocation(String url) throws IOException;
}

private static final RedirectResolver HTTP_HEAD_REDIRECT_RESOLVER = url -> {
    Head head = new Head(url);
    head.disableRedirectHandling();
    head.executeAsString();
    int status = head.getStatusCode();
    return status >= 300 && status < 400 ? head.getLocationHeader() : null;
};

String resolveShortUrl(String url, RedirectResolver resolver) {
    String current = url;
    for (int hop = 0; hop < MAX_REDIRECT_HOPS; hop++) {
        String location;
        try {
            location = resolver.resolveLocation(current);
        } catch (IOException e) {
            log.warning("Cannot resolve redirect for " + current + ": " + e);
            return null;
        }
        if (location == null)
            return current;
        current = location;
    }
    log.warning("Too many redirects resolving " + url);
    return null;
}
```
Needs `import slash.navigation.rest.Head;` and `import java.io.IOException;` (the latter likely already needed elsewhere).

In `processURL`, handle the new prefix by resolving then recursing through the normal path (add this branch before the existing `if (url.startsWith("/dir/"))`):
```java
if (url.startsWith("SHORTLINK:")) {
    String resolved = resolveShortUrl(url.substring(10), HTTP_HEAD_REDIRECT_RESOLVER);
    if (resolved == null)
        return;
    String innerFound = internalFindUrl(resolved);
    if (innerFound != null)
        processURL(innerFound, encoding, context);
    return;
}
```

**C. `GoogleMapsUrlFormat.java` — `/place/` single-location parsing.**

Add a coordinate-extraction pattern (next to the other patterns) — verified against the live capture above (`!3d<lat>!4d<lng>`, in that lat/lon order — note this is the OPPOSITE order convention from the `!1d<lon>!2d<lat>` pairs already (and deliberately) ignored inside `/dir/` waypoints, so do not generalize/share logic between the two):
```java
private static final Pattern PLACE_COORDINATE_PATTERN = Pattern.compile(".*?!3d(-?\\d+\\.\\d+)!4d(-?\\d+\\.\\d+).*");
```

Add:
```java
Wgs84Position parsePlacePosition(String url) {
    int slash = url.indexOf('/');
    String namePart = slash == -1 ? url : url.substring(0, slash);
    String name = trim(decodeUri(namePart));
    Matcher matcher = PLACE_COORDINATE_PATTERN.matcher(url);
    Double latitude = null, longitude = null;
    if (matcher.matches()) {
        latitude = parseDouble(matcher.group(1));
        longitude = parseDouble(matcher.group(2));
    }
    return asWgs84Position(longitude, latitude, name);
}
```

In `processURL`, add (alongside the `/dir/` branch, as an `else if`):
```java
else if (url.startsWith("/place/")) {
    Wgs84Position position = parsePlacePosition(url.substring(7));
    if (position != null)
        context.appendRoute(createRoute(Route, null, singletonList(position)));
}
```
(needs `import static java.util.Collections.singletonList;`)

**D. `GoogleMapsUrlFormat.java` — `?q=` parameter parsing.**

In `parsePositions(Map<String, List<String>> parameters)`, add after the existing `saddr`/`daddr`/`geocode` block, only firing when nothing else already produced a position (so it never interferes with the existing `saddr`/`daddr` tests, none of which use `q`):
```java
if (result.isEmpty()) {
    List<String> queryPositions = parameters.get("q");
    if (queryPositions != null) {
        for (String q : queryPositions) {
            result.add(parseQueryPosition(q));
        }
    }
}
```
Add:
```java
Wgs84Position parseQueryPosition(String q) {
    String trimmed = trim(q);
    if (PLAIN_POSITION_PATTERN.matcher(trimmed).matches())
        return parsePlainPosition(trimmed);
    return asWgs84Position(null, null, trimmed);
}
```
(parameter values arriving here are already percent-/`+`-decoded by `parseURLParameters`, same as `saddr`/`daddr` — no extra decoding needed, consistent with existing code.)

### Testable seam
This class has no GUI dependency already — the whole fix is directly unit-testable in `navigation-formats`, no extraction needed. Required new test cases in `GoogleMapsUrlFormatTest`:
- `testParsePlaceFromLiveCapture` — use the exact live-captured Berlin URL as a new `INPUT11_PLACE_2026` constant (`https://www.google.com/maps/place/Berlin/@52.5068767,13.4247534,10z/data=!3m1!4b1!4m6!3m5!1s0x47a84e373f035901:0x42120465b5e3b70!8m2!3d52.5200066!4d13.404954!16zL20vMDE1NnE?entry=ttu&g_ep=EgoyMDI2MDgxMi4wIKXMDSoASAFQAw%3D%3D`), assert 1 position, description `"Berlin"`, latitude ≈ 52.5200066, longitude ≈ 13.404954.
- `testParseQueryCoordinates` — `https://maps.google.com/maps?q=52.520008,13.404954`, assert 1 position, no description, lat/lng match.
- `testParseQueryPlaceName` — `https://maps.google.com/maps?q=Berlin,+Germany`, assert 1 position, description `"Berlin, Germany"`, null lat/lng.
- `testIsParseableShortLink` — `isParseableUrl("https://maps.app.goo.gl/AbCdEfGhiJ")` and `isParseableUrl("https://goo.gl/maps/AbCdEfGhiJ")` both `true`.
- `testResolveShortUrlSingleHop` — `resolveShortUrl("https://maps.app.goo.gl/x", resolver)` with a fake `RedirectResolver` returning the target URL on the first call, then `null` on the second call for that target URL → returns the target URL.
- `testResolveShortUrlMultipleHops` — fake resolver chains 3 hops before returning `null` → returns the final URL.
- `testResolveShortUrlTooManyHops` — fake resolver always returns a new location (never `null`) → returns `null` after `MAX_REDIRECT_HOPS` attempts, doesn't loop forever.
- `testResolveShortUrlIOException` — fake resolver throws `IOException` → returns `null`, no exception propagates.
- Regression: all existing `GoogleMapsUrlFormatTest` and `GoogleMapsUrlFormatBookmarkTest` cases must stay green unmodified (proves `/dir/`, `saddr`/`daddr`, bookmark `.url` handling are untouched).

### i18n
Not applicable — `GoogleMapsUrlFormat` has no dialog/UI strings; `getName()` ("Google Maps URL (*.url)") is unchanged.

### Acceptance criteria
- All new + existing tests in `GoogleMapsUrlFormatTest` and `GoogleMapsUrlFormatBookmarkTest` pass; full `mvn test` for `navigation-formats` and `rest` modules green.
- Manual maintainer check (needs live network, not CI): paste a real `https://maps.app.goo.gl/...` share link (generate one via Google Maps' own Share button) into RouteConverter and confirm a one-position route is created at the correct location.

### PR DESCRIPTION
**Pull request by cpesch**
All four files listed in the issue's "Files" section are modified, nothing else touched. Here's the summary.

## Summary

Fixed three URL shapes that `GoogleMapsUrlFormat` couldn't handle, per the maintainer's spec in issue #301:

**`rest` module** — new redirect-control API (no behavior change to existing callers):
- `HttpRequest.java`: added `protected disableRedirectHandling()`, mirroring the existing `disableContentCompression()`.
- `ReadRequest.java`: added public `disableRedirectHandling()` and `getLocationHeader()` (reusing the already-imported `HttpHeaders.LOCATION`). `Head`/`Get` inherit both for free.

**`navigation-formats` module** — `GoogleMapsUrlFormat.java`:
1. **`/place/<name>/...` single-location links** — new `PLACE_COORDINATE_PATTERN` (`!3d<lat>!4d<lng>`) and `parsePlacePosition`, wired into `isParseableUrl` and a new `/place/` branch in `processURL`. Left `/dir/` parsing untouched as instructed.
2. **`?q=<value>` parameter** — `parsePositions(Map)` now falls back to the `q` parameter (coordinate or place name) only when `saddr`/`daddr`/`geocode` produced nothing, via new `parseQueryPosition`.
3. **`maps.app.goo.gl`/`goo.gl/maps` short links** — new `SHORT_URL_PATTERN` detected in `internalFindUrl` before the existing `URL_PATTERN`, tagged `"SHORTLINK:"`, and resolved via a bounded (`MAX_REDIRECT_HOPS = 5`) `resolveShortUrl` loop using an injectable `RedirectResolver` (production impl: HTTP HEAD via the new `Head`/redirect-disable API). The resolved URL is re-run through `internalFindUrl`/`processURL`.

Added all the test cases specified (`testParsePlaceFromLiveCapture`, `testParseQueryCoordinates`, `testParseQueryPlaceName`, `testIsParseableShortLink`, and the four `resolveShortUrl` hop/exception tests) to `GoogleMapsUrlFormatTest.java`, and left every existing test untouched.

**Verify result: unavailable.** `factory-verify` returned exit 2 with a broker error (`repo not allowed (rc/* only)`), and direct `mvn` invocation was blocked by the sandbox's approval gate with no user present to approve it in this run. I compensated with a careful manual trace of every regex and control-flow path against the live-capture fixtures and the new test expectations (documented reasoning above), but this has **not** been compiled or executed — that should be double-checked in CI before merge.

Source issue: cpesch/RouteConverter#301

---
**Builder stats** · model `sonnet` · confidence `0` · 3m 45s across 1 round(s) · 4 file op(s)
[Workflow run](http://127.0.0.1:3000/cpesch/RouteConverter/actions/runs/21037)

⚠ UNVERIFIED: target not verifiable by the broker (non-rc/* target or no pom.xml — v1 is maven-only)

Closes #301

### CODE REVIEW

**Comment by cpesch** [comment]
## Verdict: request-changes · confidence 0.85

_Reviewer panel: correctness._

correctness: Short-link resolution has an unbounded recursion path when the very first HEAD response is not a 3xx (e.g. a dead, blocked, or already-expanded short link): resolveShortUrl returns the input unchanged, and processURL re-detects that unchanged string as a SHORTLINK and recurses into itself forever, each iteration issuing a live HEAD request. Everything else in this diff (place/query parsing, hop-bounded multi-redirect resolution) traces correctly against the new tests.

### navigation-formats/src/main/java/slash/navigation/url/GoogleMapsUrlFormat.java

- **major** [correctness] line 124 — If resolveShortUrl's resolver returns null on its very first call (line 148: `if (location == null) return current;`, hop 0), `resolved` equals the original short-URL text unchanged. internalFindUrl(resolved) then re-matches SHORT_URL_PATTERN and returns the identical "SHORTLINK:..." string, so `processURL(innerFound, encoding, context)` here calls itself with the exact same argument — unbounded recursion (each cycle issuing a fresh live HEAD request) ending in a StackOverflowError. This is reachable for any ordinary non-3xx HEAD response (200, 404, 403, a short link that expands via JS/meta-refresh instead of an HTTP redirect, etc.), which is a very plausible real-world response and isn't exercised by any of the new tests (they only test resolveShortUrl in isolation, never through processURL, and always either raise IOException or eventually redirect).  _(correctness)_

```suggestion
In processURL's SHORTLINK branch, detect when resolveShortUrl returned the same string it was given (no redirect occurred) and treat that as a terminal non-route rather than re-running internalFindUrl/processURL on it — e.g. compare `resolved` against `url.substring(10)` and return early if equal, or have resolveShortUrl return null explicitly for the immediate-no-redirect case so callers cannot loop back into it.
```

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher pushed a fix to `agent/builder/301-maintainer-support-current-google-maps-u`.

File: navigation-formats/src/main/java/slash/navigation/url/GoogleMapsUrlFormat.java [hunk 1]
```diff
@@ -58,6 +61,21 @@ public class GoogleMapsUrlFormat extends BaseUrlParsingFormat {
                     "(@?(" + WHITE_SPACE + "[-|\\d|\\.]+" + WHITE_SPACE + ")," +
                     "(" + WHITE_SPACE + "[-|\\d|\\.]+" + WHITE_SPACE + "))?");
     private static final String DESTINATION_SEPARATOR = "to:";
```
Comment (cpesch) [inline navigation-formats/src/main/java/slash/navigation/url/GoogleMapsUrlFormat.java:63]: **minor** [correctness] SHORT_URL_PATTERN's `[^\s]+` capture group is greedy and has no boundary other than whitespace, so any non-whitespace character directly abutting a pasted short URL (a sentence-ending period, comma, or closing parenthesis in real share text) gets absorbed into the captured URL.
```suggestion
Bound the path capture to short-link-safe characters (e.g. `[A-Za-z0-9_/-]+`) instead of `[^\s]+`, or strip trailing punctuation from the captured group before passing it to resolveShortUrl.
```

**Review by cpesch — COMMENTED** [review]
## Verdict: comment-only · confidence 0.82

_Reviewer panel: correctness._

correctness: The additive /place/, ?q=, and short-link (goo.gl) parsing paths are internally consistent with the accompanying tests and don't appear to regress the existing /dir/ or generic-URL handling. One real edge-case bug found in the short-link regex; author-flagged as unverified by compile/CI, which this review could not independently check (no repo/network access from this session).

### navigation-formats/src/main/java/slash/navigation/url/GoogleMapsUrlFormat.java

- **minor** [correctness] line 63 — SHORT_URL_PATTERN's `[^\s]+` capture group is greedy and has no boundary other than whitespace, so any non-whitespace character directly abutting a pasted short URL (a sentence-ending period, comma, or closing parenthesis in real share text) gets absorbed into the captured URL.  _(correctness)_

```suggestion
Bound the path capture to short-link-safe characters (e.g. `[A-Za-z0-9_/-]+`) instead of `[^\s]+`, or strip trailing punctuation from the captured group before passing it to resolveShortUrl.
```

---
🤖 Inspector panel — reply with `/inspector recheck` to re-run after pushing changes.

**Comment by cpesch** [comment]
🤖 Patcher **aborted** — no fix pushed.

**Reason:** agentic run made no working-tree change. Model summary: The maintainer's comment is `/inspector recheck` — a slash command directed at the Inspector minion (to re-verify/rebuild this PR), not a code change request for the Builder/Fixer minion. There is no described bug or requested edit for me to act on here.

Per instructions, when no safe fix is identifiable I should make no edits and explain why — which is what I'm doing. Working tree is already clean and matches the PR's existing commits.

No changes made: maintainer requested `/inspector recheck`, not a builder task

The feedback comment (`/inspector recheck`) is a command routed to the Inspector minion, not a change request for the Builder/Fixer. It contains no description of a bug or desired edit to `GoogleMapsUrlFormat`, `HttpRequest`, `ReadRequest`, or any other file in this PR. Since there is nothing actionable for the Builder role, no files were modified — the working tree remains clean on `agent/builder/301-maintainer-support-current-google-maps-u`. The workflow should route this comment to the Inspector minion instead.
**Confidence:** 0

Full dry-run detail on the Gitea branch `agent/patcher/dryrun-21050` (https://git.routeconverter.com/rc/rc-meta/src/branch/agent/patcher/dryrun-21050).
